removenode: allow specifying nodes to ignore using host_id

bhalevy opened the pull request:
Currently, when specifying nodes to ignore for replace or removenode,
we support specifying them only using their ip address.

As discussed in https://github.com/scylladb/scylladb/issues/11839 for removenode,
we intentionally require the host uuid for specifying the node to remove,
so the nodes to ignore (that are also done, otherwise we need not ignore them),
should be consistent with that and be specified using their host_id.

The series extends the apis and allows either the nodes ip address or their host_id
to be specified, for backward compatibility.

We should deprecate the ip address method over time and convert the tests and management
software to use the ignored nodes' host_id:s instead.

bhalevy commented:
cc @annastuchlik for the docs changes in the series.

bhalevy commented:
Note, this PR is ready for review, but I converted it to a draft since
it depends on https://github.com/scylladb/scylladb/pull/11840 which I separated out to expedite it.

asias commented on api/storage_service.cc:
```diff
@@ -717,15 +717,14 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
         auto host_id = req->get_query_param("host_id");
         std::vector<sstring> ignore_nodes_strs= split(req->get_query_param("ignore_nodes"), ",");
-        auto ignore_nodes = std::list<gms::inet_address>();
+        auto ignore_nodes = std::list<sstring>();
```
I understand why you convert to string here. But it is sad. Can we do the full parse of  the ignore_nodes here?

asias replied:
Because we know the host_id in the string is for sure host_id. But before this patch, we pass std::list<gms::inet_address> to ss:removenode(). After, we pass a std::list<sstring>. This is a step back.  We lose the type info. If after this patch, we must to do the parse in removenode, let's do all the parse there, i.e., moving the parse from ss::remove_node.set to ss:removenode. Not part of the parse in api and part in stroage_service.

asias replied:
We could verify the host id is at least a uuid in the api layer. 
But see my comments above, move all the parse stroage_sercie:: removenode().  

Do you agree it is better to do parse in one place?

bhalevy replied:
In general, I think the api layer should be responsible for the superficial parsing and validating the syntax, rejecting invalid arguments, while the semantic validation should be done at the service layer.
I.e. it's out of scope of the service::storage_service layer to make sure the list of nodes to ignore is all UUIDs or all ip addresses - that's the api's business.

We can change the storage_service api to get strongly typed parameters and do further validation based on the token_metadata, but I think it's a hassle, and in this case it's fine to pass strings down the stack.

In the next version, I'll some basic validation at the api layer so the strings are at least validated from the syntax perspective.

asias commented on locator/token_metadata.hh:
```diff
@@ -220,6 +220,10 @@ public:
 
+    /// Parses the \c host_id_string either as ip address or host uuid and returns the mapping.
+    /// Throws std::invalid_argument on parse error or std::runtime_error if the host_id wasn't found.
+    std::pair<inet_address, host_id> parse_endpoint_and_host_id(const sstring& host_id_string) const;
```
It is better to have a 
struct inet_adddres_or_host_id {
}

asias commented on service/storage_service.cc:
```diff
@@ -2341,6 +2342,11 @@ future<> storage_service::removenode(sstring host_id_string, std::list<gms::inet
             auto endpoint = *endpoint_opt;
             auto tokens = tmptr->get_tokens(endpoint);
             auto leaving_nodes = std::list<gms::inet_address>{endpoint};
+            std::list<gms::inet_address> ignore_nodes;
```
If we parse the host_id in api/, we can keep the strong type for ignore_nodes parameter.

bhalevy replied:
what about the host_id? it's also passed as a string so we should do all the parsing in the api layer or here.
I prefer to do it here since we have the token_metadata at hand, while the api layer doesn't know about it and it's current version.

asias replied:
We can convert the host_id to uuid and pass too. But  see my comments above, we can do the parse in token_metada.

asias commented on api/api-doc/storage_service.json:
```diff
@@ -1245,7 +1245,7 @@
                   {
                      "name":"ignore_nodes",
-                     "description":"List of dead nodes to ingore in removenode operation",
+                     "description":"Comma-separated list of dead nodes (host_ids or ip addresses) to ingore in removenode operation",
```
What if user passes a mix of host id and ip address, can we detect and reject this?

bhalevy replied:
Why do we care?
I think that as a follow-up we should deprecate the use of ip-addresses for ignore nodes, and eventually we'll support only host IDs.

asias replied:
This is a user api change. We care because we need to make sure wrong user inputs are handled. Do we handle it correct?

asias replied:
Mixing the host id and ip are most probably an operation mistake. I think it is better just to reject such user input. The list has to be either all host id or all ips.

asias commented:
Benny, I like what is achieved in this PR. Let's figure out some thing better than mixing host id an ip address in string though.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2776/

bhalevy commented:
In v2 ([c8886e1](https://github.com/scylladb/scylladb/commit/c8886e1fce4008f0b70b20ed71a25f033d14580a)):
- define and use locator::host_id_or_endpoint
- validate remove_node parameters syntax in api layer
  - pass strongly typed params to storage_service::removenode
  - do the semantic validation at the service layer by resolving hsot_id to endpoint or vice versa

asias commented:
v2 looks good. Thanks.

denesb commented:
@bhalevy it looks like the dependency has been merged. Please rebase and take it out of Draft if it is ready for merging.

denesb commented on locator/token_metadata.hh:
```diff
@@ -133,6 +133,27 @@ private:
     bool _sort_by_proximity = true;
 };
 
+struct host_id_or_endpoint {
```
This should maybe be in api/, if the only user of it is there.

bhalevy replied:
storage_service needs it to.
I think that since the host id and endpoint are location information this structure should be in the locator namespace. maybe in a separate header file, but we can extract it later to.

In the future, when we use either the endpoint or the host_id to identify nodes, we'll need a registry and a `class node` abstraction.  The mapping registry today is effectively held in the token metadata.

bhalevy replied:
Maybe we should extract stuff to locator/topology.hh as a follow up.

denesb replied:
If it has more users, I'm fine with it staying here.

denesb commented on api/storage_service.cc:
```diff
@@ -722,18 +722,26 @@ void set_storage_service(http_context& ctx, routes& r, sharded<service::storage_
+                        if (hoep.has_host_id() && !ignore_nodes.front().has_host_id()) {
+                            throw std::runtime_error("Expected host UUID. Use the same method for all nodes to ignore: either Host IDs or ip addresses.");
+                        }
+                        if (hoep.has_endpoint() && !ignore_nodes.front().has_endpoint()) {
```
This can be condensed into a single if:
```
if (hoep.has_host_id() != ignore.nodes.front().has_host_id()) {
```

bhalevy replied:
I wanted the error to be more user friendly about what is expected, but maybe it's not that important, after all the list should be symmetrical.

bhalevy replied:
done in v3

denesb commented on api/api-doc/storage_service.json:
```diff
@@ -1228,7 +1228,7 @@
             {
                "method":"POST",
-               "summary":"Removes token (and all data associated with enpoint that had it) from the ring",
+               "summary":"Removes a node from the cluster. Replicated data that logically belonged to this node is remapped and restored onto the remaining nodes, if possible.",
```
s/remapped and restored onto/redistributed among/

Why the `if possible`? Is there a situation under this is dong possible? A failure is always implicit I think, no need to include it in the message.

bhalevy replied:
I think we can be in a degraded situation where we cannot satisfy the replication factor if removing a node when RF == number of nodes.

bhalevy replied:
done in v3

bhalevy commented:
In v3 [335a8cc](https://github.com/scylladb/scylladb/commit/335a8cc36286411f224228cd441b865b22ea9f01)):
- rebased + retested
- simplified `hoep.has_host_id() != ignore.nodes.front().has_host_id()` condition and error message
- `s/remapped and restored onto/redistributed among/` in remove_node json description

denesb reviewed (approved):
LGTM, will queue once CI approves.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2789/
